### PR TITLE
Handle Next.js rewrites/redirects/headers

### PR DESCRIPTION
**Pull request by leoortizz**
### Description
- Handle Next.js rewrites/redirects/headers incompatible with `firebase.json` in Cloud Functions
- Filter out Next.js prerendered routes that matches rewrites/redirects/headers rules from SSG content directory

### CODE REVIEW

File: src/frameworks/next/index.ts [hunk 1]
```diff
@@ -149,11 +150,25 @@ export async function build(dir: string): Promise<BuildResult> {
...
+      if (
+        has ||
+        destination.includes("http") // filter out rewrites to third parties
+      )
+        return undefined;
+
+      try {
+        new RE2(source);
+        new RE2(destination);
+
```
Comment (jamesdaniels) [inline src/frameworks/next/index.ts:166]: Set `wantsBackend` to true

File: src/frameworks/next/index.ts [hunk 2]
```diff
@@ -149,11 +150,25 @@ export async function build(dir: string): Promise<BuildResult> {
...
+        return { source, destination };
+      } catch (error) {
+        // regex not supported by RE2 cannot be transformed into firebase.json rewrites
+        return undefined;
```
Comment (jamesdaniels) [inline src/frameworks/next/index.ts:170]: Set `wantsBackend` to true

File: src/frameworks/next/index.ts [hunk 3]
```diff
@@ -149,11 +150,25 @@ export async function build(dir: string): Promise<BuildResult> {
...
-      if (has) return undefined;
-      return { source, destination };
+      if (
+        has ||
+        destination.includes("http") // filter out rewrites to third parties
+      )
+        return undefined;
+
+      try {
+        new RE2(source);
```
Comment (jamesdaniels) [inline src/frameworks/next/index.ts:164]: let's do similar tests for redirects and headers

File: src/frameworks/utils.ts [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+  } catch (error) {
+    return false;
+  }
+}
+
+/**
+ * Wheter the given URL is an external URL.
+ */
+export function isThirdPartyUrl(url: string): boolean {
+  return url.includes("http");
```
Comment (jamesdaniels) [inline src/frameworks/utils.ts:21]: I'd do `.startsWith` and change it to just isUrl

File: src/frameworks/next/utils.ts [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+ * According to the Next.js documentation:
+ * ```md
+ *  To match a regex path you can wrap the regex in parentheses
+ *  after a parameter, for example /post/:slug(\\d{1,}) will match /post/123
+ *  but not /post/abc.
+ * ```
+ * See: https://nextjs.org/docs/api-reference/next.config.js/redirects#regex-path-matching
+ */
+export function pathHasRegex(path: string): boolean {
+  for (let i = 0; i < path.length; i++) {
```
Comment (jamesdaniels) [inline src/frameworks/next/utils.ts:12]: This would probably be more readable, tunable with a regex, mind converting?

File: src/frameworks/next/index.ts [hunk 6]
```diff
@@ -177,18 +187,18 @@ export async function build(dir: string): Promise<BuildResult> {
...
       if (has) return false;
 
-      if (
-        supportsFrameworkRegex(source) &&
-        supportsFrameworkRegex(destination) &&
-        isUrl(destination) === false
-      ) {
-        return true;
-      } else {
+      if (pathHasRegex(source) || isUrl(destination)) {
```
Comment (jamesdaniels) [inline src/frameworks/next/index.ts:190]: Maybe we could break these into util functions? `filter(rewriteSupportedByFirebase)` that would make them easier to test & reuse in codegenPublicDir too.

would you mind taking on the work to utilize in codegenPublicDir in this PR? the idea being if the regex matches the path (for !supportedByFirebase) we should not copy that file into the CDN

**Review by jamesdaniels — CHANGES_REQUESTED** [review]

**Comment by jamesdaniels** [comment]
Add changelog

File: src/frameworks/next/utils.ts (hunk 5 above)
Comment (leoortizz) [reply src/frameworks/next/utils.ts]: @TheIronDev in the last commit I pushed a working regex for that.. I think it can get tricky with `.includes` because of paths like this:
```
/post/\\(escapedparentheses\\)/:slug(\\d{1,}) <--- this has a regex and escaped parenthesis
```

File: src/frameworks/next/utils.ts [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+  }
+
+  return false;
+}
+
+/**
+ * Remove double backslashes from a string
+ */
+export function cleanEscapedChars(path: string): string {
+  return path.replace(/\\/g, "");
```
Comment (jamesdaniels) [inline src/frameworks/next/utils.ts:29]: Shouldn't this replace with a single "\", rather than empty? mind getting some more edge cases covered in the test?

File: src/frameworks/next/utils.ts [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+ * ```md
+ *  To match a regex path you can wrap the regex in parentheses
+ *  after a parameter, for example /post/:slug(\\d{1,}) will match /post/123
+ *  but not /post/abc.
+ * ```
+ * See: https://nextjs.org/docs/api-reference/next.config.js/redirects#regex-path-matching
+ */
+export function pathHasRegex(path: string): boolean {
+  // finds parentheses that are not preceded by double backslashes
+  const regex = /(?<!\\)\(/;
```
Comment (jamesdaniels) [inline src/frameworks/next/utils.ts:16]: I don't think it would hurt readability to make this a one liner, e.g, `return /(?<!\\)\(/.test(path)`

File: src/frameworks/utils.ts [hunk 9]
```diff
@@ -0,0 +1,6 @@
+/**
+ * Whether the given string is a URL.
+ */
+export function isUrl(url: string): boolean {
+  return url.startsWith("http");
```
Comment (jamesdaniels) [inline src/frameworks/utils.ts:5]: we should make sure this is a protocol, /^https?:\/\//, so say `httpendpoint/foo/bar` wouldn't trip it

**Review by jamesdaniels — CHANGES_REQUESTED** [review]

File: src/frameworks/utils.ts (hunk 9 above)
Comment (leoortizz) [reply src/frameworks/utils.ts]: @jamesdaniels 🤦‍♂️ right, thanks! I also added the example you mentioned as test cases

File: src/frameworks/next/utils.ts (hunk 7 above)
Comment (leoortizz) [reply src/frameworks/next/utils.ts:23]: @jamesdaniels right.. that should be the case. Reviewing it now Next.js has this on their docs:

> The following characters (, ), {, }, :, *, +, ? are used for regex path matching, so when used in the source as non-special values they must be escaped by adding

I'm fixing it and adding test cases, will push

File: src/frameworks/next/interfaces.ts [hunk 10]
```diff
@@ -0,0 +1,31 @@
+import type { Header, Rewrite, Redirect } from "next/dist/lib/load-custom-routes";
+
```
Comment (leoortizz) [inline src/frameworks/next/interfaces.ts:2]: Unfortunately `Redirect` is not extendable because its members are not statically known, so this is not possible:
```
export interface RoutesManifestRedirect extends Redirect {
  regex: string;
}
```

File: src/frameworks/next/index.ts [hunk 11]
```diff
@@ -136,26 +127,71 @@ export async function build(dir: string): Promise<BuildResult> {
...
+  const manifest = JSON.parse(manifestBuffer.toString()) as Manifest;
   const {
     headers: nextJsHeaders = [],
     redirects: nextJsRedirects = [],
     rewrites: nextJsRewrites = [],
   } = manifest;
-  const headers = nextJsHeaders.map(({ source, headers }) => ({ source, headers }));
+
+  const headers = nextJsHeaders
+    .filter((header) => {
```
Comment (TheIronDev) [inline src/frameworks/next/index.ts:138]: Nit: Please remove side-effects from the filter function.

Maybe instead:

```
const hasUnsupportedHeader = nextJsHeaders.some((header) => {
  return !isHeaderSupportedByFirebase(header);
});
if (hasUnsupportedHeader) {
 wantsBackend = true;
}

const headers = nextJsHeaders
  .filter((header) => isHeaderSupportedByFirebase(header))
  .map(({ source, headers }) => ({
      // clean up unnecessary escaping
      source: cleanEscapedChars(source),
      headers,
    }));
```

File: src/frameworks/next/index.ts [hunk 12]
```diff
@@ -136,26 +127,71 @@ export async function build(dir: string): Promise<BuildResult> {
...
-  const nextJsRewritesToUse = Array.isArray(nextJsRewrites)
-    ? nextJsRewrites
-    : nextJsRewrites.beforeFiles || [];
+    .filter((redirect) => {
+      if (redirect.internal) return false;
+
+      if (isRedirectSupportedByFirebase(redirect)) {
+        return true;
+      } else {
+        wantsBackend = true;
```
Comment (TheIronDev) [inline src/frameworks/next/index.ts:159]: Same here

File: src/frameworks/next/index.ts [hunk 13]
```diff
@@ -136,26 +127,71 @@ export async function build(dir: string): Promise<BuildResult> {
...
   const rewrites = nextJsRewritesToUse
-    .map(({ source, destination, has }) => {
-      // Can we change i18n into Firebase settings?
-      if (has) return undefined;
-      return { source, destination };
+    .filter((rewrite) => {
+      if (isRewriteSupportedByFirebase(rewrite)) {
+        return true;
+      } else {
+        wantsBackend = true;
```
Comment (TheIronDev) [inline src/frameworks/next/index.ts:186]: Ditto here

File: src/frameworks/next/index.ts [hunk 14]
```diff
@@ -210,10 +246,37 @@ export async function ɵcodegenPublicDirectory(sourceDir: string, destDir: strin
...
+        join(
+          sourceDir,
+          distDir,
+          "routes-manifest.json" // TODO: get this from next/constants
+        )
+      ),
+    ]);
+
     const prerenderManifest = JSON.parse(prerenderManifestBuffer.toString());
+    const routesManifest = JSON.parse(routesManifestBuffer.toString()) as Manifest;
```
Comment (TheIronDev) [inline src/frameworks/next/index.ts:267]: Is it possible `routesManifestBuffer` or `prerenderManifestBuffer` is not JSON.parse-able?

Should we try-catch guard this to prevent errors?

File: src/frameworks/next/index.ts [hunk 15]
```diff
@@ -210,10 +246,37 @@ export async function ɵcodegenPublicDirectory(sourceDir: string, destDir: strin
...
+
+    const rewritesToUse = getNextjsRewritesToUse(routesManifest.rewrites);
+    const rewritesNotSupportedByFirebase = rewritesToUse?.filter(
+      (rewrite) => isRewriteSupportedByFirebase(rewrite) === false
```
Comment (TheIronDev) [inline src/frameworks/next/index.ts:271]: Nit, these can be simplified:

```
(rewrite) => !isRewriteSupportedByFirebase(rewrite)
```

File: src/frameworks/next/index.ts [hunk 16]
```diff
@@ -210,10 +246,37 @@ export async function ɵcodegenPublicDirectory(sourceDir: string, destDir: strin
...
+    );
+    const redirectsNotSupportedByFirebase = routesManifest.redirects?.filter(
+      (redirect) => isRedirectSupportedByFirebase(redirect) === false
+    );
+    const headersNotSupportedByFirebase = routesManifest.headers?.filter(
```
Comment (TheIronDev) [inline src/frameworks/next/index.ts:276]: If you guard with an empty array you can remove a bunch of the if-blocks later on.

```
const headersNotSupportedByFirebase = routesManifest.headers?.filter(
      (header) => isHeaderSupportedByFirebase(header) === false
    ) || [];


```

File: src/frameworks/next/interfaces.ts (hunk 10 above)
Comment (TheIronDev) [reply src/frameworks/next/interfaces.ts:2]: :(

File: src/frameworks/next/utils.ts [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+ */
+export function cleanEscapedChars(path: string): string {
+  return path.replace(/\\/g, "");
+}
+
+/**
+ * Whether a Next.js rewrite is supported by Firebase.
+ */
+export function isRewriteSupportedByFirebase(rewrite: Rewrite): boolean {
+  if ("has" in rewrite || pathHasRegex(rewrite.source) || isUrl(rewrite.destination)) {
```
Comment (TheIronDev) [inline src/frameworks/next/utils.ts:31]: Simplify Nit:

```
return !("has" in rewrite || pathHasRegex(rewrite.source) || isUrl(rewrite.destination));
```

Comment (TheIronDev) [reply src/frameworks/next/utils.ts:31]: + others

Comment (TheIronDev) [inline src/frameworks/next/utils.ts:31]: What is the significance of "has" here? Can you add a comment explaining how this is being checked?

File: src/frameworks/next/utils.ts [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+  return true;
+}
+
+/**
+ * Firebase Rewrites that can be supported by firebase.json.
+ */
+export function getNextjsRewritesToUse(
+  nextJsRewrites: Manifest["rewrites"]
+): RoutesManifestRewrite[] {
+  return Array.isArray(nextJsRewrites) ? nextJsRewrites : nextJsRewrites?.beforeFiles || [];
```
Comment (TheIronDev) [inline src/frameworks/next/utils.ts:66]: I see 3 possible return values packed into 1 line.
Can we break up this ternary to make it easier to read?

```
if (Array.isArray(nextJsRewrites)) {
  return nextJsRewrites;
}

return nextJsRewrites?.beforeFiles || [];
```

Also, are we certain `nextJsRewrites?.beforeFiles` is an array? should we instead check `Array.isArray` on it to be safe?

File: src/frameworks/next/index.ts (hunk 14 above)
Comment (jamesdaniels) [reply src/frameworks/next/index.ts:267]: probably better to refactor to readJSON & throw rather than get too crazy IMO

File: src/frameworks/next/index.ts [hunk 19]
```diff
@@ -210,10 +246,37 @@ export async function ɵcodegenPublicDirectory(sourceDir: string, destDir: strin
...
+    ]);
+
     const prerenderManifest = JSON.parse(prerenderManifestBuffer.toString());
+    const routesManifest = JSON.parse(routesManifestBuffer.toString()) as Manifest;
+
+    const rewritesToUse = getNextjsRewritesToUse(routesManifest.rewrites);
+    const rewritesNotSupportedByFirebase = rewritesToUse?.filter(
+      (rewrite) => isRewriteSupportedByFirebase(rewrite) === false
+    );
+    const redirectsNotSupportedByFirebase = routesManifest.redirects?.filter(
```
Comment (jamesdaniels) [inline src/frameworks/next/index.ts:273]: Can we destructure with defaults, `const { redirects=[] } = routesManifest` like we do in build? would make it so this couldn't be null

Comment (jamesdaniels) [reply src/frameworks/next/index.ts:273]: and we wouldn't have to guard

File: src/frameworks/next/index.ts [hunk 20]
```diff
@@ -222,6 +285,29 @@ export async function ɵcodegenPublicDirectory(sourceDir: string, destDir: strin
           continue;
         }
 
+        if (rewritesNotSupportedByFirebase) {
```
Comment (jamesdaniels) [inline src/frameworks/next/index.ts:288]: If we destructure / guard, then we can simplify this (and the others) as

```ts
if (rewritesNotSupportedByFirebase.find(rewrite => new RegExp(rewrite.regex).test(path))) {
  continue;
}```

**Review by jamesdaniels — CHANGES_REQUESTED** [review]

File: src/frameworks/next/index.ts (hunk 20 above)
Comment (jamesdaniels) [reply src/frameworks/next/index.ts:288]: As an optimization you would pre init all the regex before the `for`, `e.g, const rewriteRegexsNotSupportedByFirebase = rewritesNotSupportedByFirebase.map(rewrite => new RegExp(rewrite.regex))`

File: src/frameworks/next/utils.ts (hunk 18 above)
Comment (leoortizz) [reply src/frameworks/next/utils.ts]: @TheIronDev [Next.js won't build](https://nextjs.org/docs/messages/routes-must-be-array) if this is not an array(or object with arrays). Since we always use this after the build step we should not need to check if it's an array

**Review by jamesdaniels — CHANGES_REQUESTED** [review]

File: src/frameworks/next/index.ts (hunk 20 above)
Comment (leoortizz) [reply src/frameworks/next/index.ts]: @jamesdaniels I changed that to destructure with defaults and also instantiated the regexes before the loop as suggested.

Then instead of
```js
if (rewritesNotSupportedByFirebase.find(rewrite => new RegExp(rewrite.regex).test(path))) {
  continue;
}
```

I assigned the loop result to a variable because I think it's easier/quicker to understand as the variable name replaces a comment that may become outdated if this changes at some point. Also as the name is quite long, IMO a variable name can help. Let me know your thoughts

File: src/frameworks/next/utils.ts (hunk 17 above)
Comment (leoortizz) [reply src/frameworks/next/utils.ts]: @TheIronDev `has` is used for [Header, Cookie, and Query Matching](https://nextjs.org/docs/api-reference/next.config.js/rewrites#header-cookie-and-query-matching). As it can't be translated to firebase.json we say it's not supported. I improved the functions docs linking to both the `firebase.json` and Next.js docs

File: src/frameworks/next/index.ts [hunk 21]
```diff
@@ -230,15 +231,15 @@ export async function ɵcodegenPublicDirectory(sourceDir: string, destDir: strin
       }
     }
 
-    const [prerenderManifestBuffer, routesManifestBuffer] = await Promise.all([
-      readFile(
+    const [prerenderManifest, routesManifest] = await Promise.all([
+      readJSON<any>(
```
Comment (jamesdaniels) [inline src/frameworks/next/index.ts:235]: `<any>` is implied

File: src/frameworks/next/index.ts [hunk 22]
```diff
@@ -136,26 +128,55 @@ export async function build(dir: string): Promise<BuildResult> {
   }
 
   const manifestBuffer = await readFile(join(dir, distDir, "routes-manifest.json"));
-  const manifest: Manifest = JSON.parse(manifestBuffer.toString());
+  const manifest = JSON.parse(manifestBuffer.toString()) as Manifest;
```
Comment (jamesdaniels) [inline src/frameworks/next/index.ts:131]: I directed incorrectly, `manifest: Manifest` is still preferred over `as Manifest` at Google

**Review by jamesdaniels — APPROVED** [review]
Address nits if you have the chance, I'll kick off integration tests

File: src/frameworks/next/utils.ts (hunk 7 above)
Comment (jamesdaniels) [reply src/frameworks/next/utils.ts:24]: @leoortizz `path.replace(/\\([\(\){}\:\+\?])/g, (a,b) => b)` should do it IMO. WDYT?

```
> '/\\(\\)\\{\\}\\:\\+\\?/:slug'.replace(/\\([\(\){}\:\+\?])/g, (a,b) => b)
'/(){}:+?/:slug'
```

File: src/frameworks/next/index.ts (hunk 22 above)
Comment (leoortizz) [reply src/frameworks/next/index.ts]: @jamesdaniels I replaced it with the typed `readJSON`

File: src/frameworks/next/index.ts (hunk 21 above)
Comment (leoortizz) [reply src/frameworks/next/index.ts]: @jamesdaniels `any` was not implied as the function required a type. As it might be necessary sometimes I added `any` as a default

File: src/frameworks/next/utils.ts (hunk 7 above)
Comment (leoortizz) [reply src/frameworks/next/utils.ts]: @jamesdaniels this seem to fix it, thanks! I implemented it and adjusted the unit tests, should be all working now

**Review by jamesdaniels — APPROVED** [review]

**Comment by jamesdaniels** [comment]
merged in #5316, thanks @leoortizz
